tapfreighter/wallet.go

```diff
@@ -511,6 +521,141 @@ func (f *AssetWallet) FundPacket(ctx context.Context,
+
+	// We need to find a commitment that has enough assets to satisfy this
+	// send request. We'll map the address to a set of constraints, so we
+	// can use that to do Taproot asset coin selection.
+	constraints := CommitmentConstraints{
+		GroupKey: fundDesc.GroupKey,
+		AssetID:  &fundDesc.ID,
+		MinAmt:   fundDesc.Amount,
+	}
+	selectedCommitments, err := f.cfg.CoinSelector.SelectCoins(
+		ctx, constraints, PreferMaxAmount,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// If we return with an error, we want to release the coins we've
+	// selected.
+	success := false
+	defer func() {
+		if !success {
+			outpoints := fn.Map(
+				selectedCommitments,
+				func(c *AnchoredCommitment) wire.OutPoint {
+					return c.AnchorPoint
+				},
+			)
+			err := f.cfg.CoinSelector.ReleaseCoins(
+				ctx, outpoints...,
+			)
+			if err != nil {
+				log.Errorf("Unable to release coins: %v", err)
+			}
+		}
+	}()
+
+	activeAssets := fn.Filter(
+		selectedCommitments, func(c *AnchoredCommitment) bool {
+			return c.Asset.ID() == fundDesc.ID
+		},
+	)
+
+	// Now that we know what inputs we're going to spend, we know that by
+	// definition, we use the first input's info as the burn's PrevID.
+	firstInput := activeAssets[0]
+	firstPrevID := asset.PrevID{
+		OutPoint: firstInput.AnchorPoint,
+		ID:       firstInput.Asset.ID(),
+		ScriptKey: asset.ToSerialized(
+			firstInput.Asset.ScriptKey.PubKey,
+		),
+	}
+	burnKey := asset.NewScriptKey(asset.DeriveBurnKey(firstPrevID))
+	newInternalKey, err := f.cfg.KeyRing.DeriveNextKey(
+		ctx, asset.TaprootAssetsKeyFamily,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// We want both the burn output and the change to be in the same anchor
+	// output, that's why we create the packet manually.
+	vPkt := &tappsbt.VPacket{
+		Inputs: []*tappsbt.VInput{{
+			PrevID: asset.PrevID{
+				ID: fundDesc.ID,
+			},
+		}},
+		Outputs: []*tappsbt.VOutput{{
+			Amount:            0,
+			Type:              tappsbt.TypeSplitRoot,
+			AnchorOutputIndex: 0,
+			ScriptKey:         asset.NUMSScriptKey,
+		}, {
+			Amount:            fundDesc.Amount,
+			Type:              tappsbt.TypeSimple,
+			Interactive:       true,
+			AnchorOutputIndex: 0,
+			ScriptKey:         burnKey,
+		}},
+		ChainParams: f.cfg.ChainParams,
+	}
+	vPkt.Outputs[0].SetAnchorInternalKey(
+		newInternalKey, f.cfg.ChainParams.HDCoinType,
+	)
+	vPkt.Outputs[1].SetAnchorInternalKey(
+		newInternalKey, f.cfg.ChainParams.HDCoinType,
+	)
+
+	// The virtual transaction is now ready to be further enriched with the
+	// split commitment and other data.
+	fundedPkt, err := f.fundPacketWithInputs(
+		ctx, fundDesc, vPkt, selectedCommitments,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// We want to avoid a BTC output being created that just sits there
+	// without an actual commitment in it. So if we are not getting any
+	// change or passive assets in this output, we'll not want to go through
+	// with it.
+	changeOut := fundedPkt.VPacket.Outputs[0]
```

Comment:
If the above is unchanged though, then doesn't this lead to us just sending those other funds to the NUMs key?

Reply:
Yes, but I think from `fundPacketWithInputs` and the logic right after this, that key will never stay as NUMS if the amount is non-zero; this either becomes a normal split, or the tombstone is removed.

Reply:
Correct. We might want to eventually refactor the whole wallet code a bit to either make some of the assumptions less implicit or by just documenting them better.